apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl

```diff
@@ -0,0 +1,379 @@
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_gateway_authn_http_SUITE).
+
+-compile(nowarn_export_all).
+-compile(export_all).
+
+-include("emqx_authn.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("common_test/include/ct.hrl").
+-include_lib("emqx/include/emqx_placeholder.hrl").
+
+-define(PATH, [?CONF_NS_ATOM]).
+
+-define(HTTP_PORT, 33333).
+-define(HTTP_PATH, "/auth").
+
+-define(checkMatch(Guard),
+    (fun(Expr) ->
+        case (Expr) of
+            Guard ->
+                ok;
+            X__V ->
+                erlang:error(
+                    {assertMatch, [
+                        {module, ?MODULE},
+                        {line, ?LINE},
+                        {expression, (??Expr)},
+                        {pattern, (??Guard)},
+                        {value, X__V}
+                    ]}
+                )
+        end
+    end)
+).
+-define(FUNCTOR(Expr), fun() -> Expr end).
+-define(FUNCTOR(Arg, Expr), fun(Arg) -> Expr end).
+
+-define(PROTOCOLS, [coap, lwm2m, 'mqtt-sn', stomp, exproto]).
+
+-define(CONFS, [
+    emqx_coap_SUITE,
+    emqx_lwm2m_SUITE,
+    emqx_sn_protocol_SUITE,
+    emqx_stomp_SUITE,
+    emqx_exproto_SUITE
+]).
+
+-define(CASES, [
+    fun case_coap/0,
+    fun case_lwm2m/0,
+    fun case_emqx_sn/0,
+    fun case_stomp/0,
+    fun case_exproto/0
+]).
+
+-define(AUTHNS, [fun set_http_authn/1]).
+
+-type auth_controller() :: fun((start | stop) -> ok).
+
+all() ->
+    emqx_common_test_helpers:all(?MODULE).
+
+init_per_suite(Config) ->
+    _ = application:load(emqx_conf),
+    ok = emqx_common_test_helpers:load_config(emqx_gateway_schema, init_conf()),
+    emqx_common_test_helpers:start_apps([emqx_authn, emqx_gateway]),
+    application:ensure_all_started(cowboy),
+    Config.
+
+end_per_suite(_) ->
+    clear_authn(),
+    ok = emqx_common_test_helpers:load_config(emqx_gateway_schema, <<>>),
+    emqx_common_test_helpers:stop_apps([emqx_authn, emqx_gateway]),
+    application:stop(cowboy),
+    ok.
+
+init_per_testcase(_Case, Config) ->
+    {ok, _} = emqx_cluster_rpc:start_link(node(), emqx_cluster_rpc, 1000),
+    Config.
+
+end_per_testcase(_Case, Config) ->
+    Config.
+
+%%------------------------------------------------------------------------------
+%% Tests
+%%------------------------------------------------------------------------------
+t_authn(_) ->
+    test_gateway_with_auths(?CASES, ?AUTHNS).
```

Comment:
Can we put the logic for starting/stopping the gateway and deps resources into another module? In that way, we only need to focus on how to organize the tests in `*_SUITE.erl`. In the _SUITE.erl, It is preferred:

```erlang

groups() ->
  emqx_gateway_ct:init_per_groups(matrix()).

matrix() ->
  %% Gateways :: [lwm2m, mqttsn, stomp, ...] | all
  %% Authns :: [auth_http, built_in_database, ...]
  %% Authzs :: [auth_http, built_in_database, ...]
  {GatewayNames, AuthnNames, AuthzNames}.

init_per_group(GrpName, _Conf) ->
  emqx_gateway_ct:init_per_group(GrpName, fun prepare/1).

end_per_group(GrpName, _Conf) ->
   emqx_gateway_ct:end_per_group(GrpName, fun cleanup/1).

prepare({GatewayName, AuthnName, AuthzName) ->
    _ = start_authn_server(AuthnName),
    _ = start_authz_server(AuthzName),
   emqx_gateway_ct:merge_confs(
    gateway_confs(GatewayName),
    authn_confs(AuthnName),
    authz_confs(AuthzName)).

cleanup({GatewayName, AuthnName, AuthzName) ->
  _ = clean_authn_resources(AuthnName),
  _ = clean_authz_resources(AuthzName),
  ok.

%% cases

t_connect_successfully(Conf) ->
    GwName = emqx_gateway_ct:gateway(Conf),
   emqx_gateway_ct:with_client(GwName, fun(C) ->
       ok
   end).

...

```